File: atcoder/abc051/d.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
#define rep(i,n) for(int (i)=0;(i)<(int)(n);++(i))
#define each(itr,c) for(__typeof(c.begin()) itr=c.begin(); itr!=c.end(); ++itr)
#define all(x) (x).begin(),(x).end()
#define pb push_back
#define fi first
#define se second

typedef pair<int,int> pi;

const int N=100;
const int INF=123456789;
struct edge{ int to,cost; };
vector<edge> G[N];
int d[N],par[N];

int main()
{
    int n,m;
    scanf(" %d %d", &n, &m);

    set<pi> used;
    rep(i,m)
    {
        int a,b,c;
        scanf(" %d %d %d", &a, &b, &c);
        --a;
        --b;
        G[a].pb(edge{b,c});
        G[b].pb(edge{a,c});
    }

    rep(i,n)
    {
        fill(d,d+N,INF);
        fill(par,par+N,-1);

        // dijkstra
        priority_queue<pi,vector<pi>,greater<pi>> que;
        d[i]=0;
        que.push(pi(0,i));
        while(!que.empty()){
            pi p=que.top();
            que.pop();
            int v=p.se;
            if(d[v]<p.fi) continue;
            rep(j,G[v].size())
            {
                edge e=G[v][j];
                if(d[e.to]>d[v]+e.cost)
                {
                    d[e.to]=d[v]+e.cost;
                    que.push(pi(d[e.to],e.to));
                    par[e.to]=v;
                }
            }
        }

        rep(j,n)if(i!=j)
        {
            int a=j, b=par[j];
            if(a>b) swap(a,b);
            used.insert(pi(a,b));
        }
    }

    cout << m - used.size() << endl;
    return 0;
}

File: atcoder/abc051/d.py
import queue

INF = 123456789

def main():
    n,m = map(int,input().split())

    G = [[] for _ in range(n)]
    for _ in range(m):
        a,b,c = map(int,input().split())
        a -= 1
        b -= 1
        G[a].append((b,c))
        G[b].append((a,c))

    used = set()
    for i in range(n):
        d = [INF for _ in range(n)]
        par = [-1 for _ in range(n)]

        # dijkstra
        que = queue.PriorityQueue()
        d[i] = 0
        que.put((0,i))
        while not que.empty():
            p = que.get()
            v = p[1]
            if d[v] < p[0]:
                continue
            for (to,cost) in G[v]:
                if d[to] > d[v]+cost:
                    d[to] = d[v]+cost
                    par[to] = v
                    que.put((d[to],to))

        for j in range(n):
            if i != j:
                a,b = j,par[j]
                if a>b:
                    a,b = b,a
                used.add((a,b))

    print(m - len(used))

if __name__ == '__main__':
    main()
